### PR TITLE
[amgcl] Fixing rigid body modes construction for elasticity problems

### PR DESCRIPTION
**Pull request by ddemidov**
This is a two-part PR:

1. Brings bug fixes from upstream, mostly for this issue: https://github.com/ddemidov/amgcl/issues/135
2. It seems the construction of near-null-space vectors (rigid body modes) in Kratos wrapper code was wrong (probably brought over from subdomain deflation code):

https://github.com/KratosMultiphysics/Kratos/blob/b176b910ff4f562932cf4bb1cb57f2300bf958b8/kratos/linear_solvers/amgcl_solver.h#L385-L397

It is shown in https://github.com/ddemidov/amgcl/issues/135 that using the correct rigid body modes with elasticity problems may reduce the number of iterations by a factor of two (compared with block-wise coarsening).

The changes here will only work for pure elasticity problems (without the pressure) in 2D or 3D. This requires some testing (I don't know how to do this myself) to see if it does actually improves convergence.

### CODE REVIEW

File: kratos/linear_solvers/amgcl_solver.h
Comment (ddemidov) [inline kratos/linear_solvers/amgcl_solver.h]: The loops here could benefit from OpenMP pragmas, not sure what is the Kratos policy regarding this.

**Comment by ddemidov** [comment]
I've pushed more upstream fixes to the PR. Specifically,

- support for mixed precision solution in mpi with vexcl and block values (https://github.com/ddemidov/amgcl/issues/137)
- Fix for the issue discovered by @RiccardoRossi and @roigcarlo where MPI solve hanged if one of the MPI domains was empty.

**Comment by RiccardoRossi** [comment]
hi denis i am getting a compile error

~~~
/home/riccardo/Kratos/applications/TrilinosApplication/external_includes/amgcl_mpi_solver_impl.cpp:15:10: fatal error: amgcl/mpi/solver/runtime.hpp: No such file or directory
 #include <amgcl/mpi/solver/runtime.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
applications/TrilinosApplication/CMakeFiles/KratosTrilinosCore.dir/build.make:86: recipe for target 'applications/TrilinosApplication/CMakeFiles/KratosTrilinosCore.dir/external_includes/amgcl_mpi_solver_impl.cpp.o' failed
make[2]: *** [applications/TrilinosApplication/CMakeFiles/KratosTrilinosCore.dir/external_includes/amgcl_mpi_solver_impl.cpp.o] Error 1
CMakeFiles/Makefile2:786: recipe for target 'applications/TrilinosApplication/CMakeFiles/KratosTrilinosCore.dir/all' failed
make[1]: *** [applications/TrilinosApplication/CMakeFiles/KratosTrilinosCore.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
~~~

**Comment by ddemidov** [comment]
Should be fixed now. Not sure if failing Ubuntu-CI tests are an issue?

**Comment by roigcarlo** [comment]
> Not sure if failing Ubuntu-CI tests are an issue?

Don't worry is a known issue, if travis passes is ok

**Comment by RiccardoRossi** [comment]
ok, it now compiles, and runs at least in my case

**Comment by RiccardoRossi** [comment]
so, i tried this out.

good thing is that it works (also in mpi), with coordinates.

bad thing is that i get the following:

~~~
{
    "precond": {
        "class": "amg",
        "relax": {
            "type": "ilu0"
        },
        "coarsening": {
            "type": "smoothed_aggregation",
            "aggr": {
                "eps_strong": "0",
                "block_size": "3"
            },
            "nullspace": {
                "cols": "6",
                "rows": "15120",
                "B": "0x37a5ff0"
            }
        },
        "npre": "1",
        "npost": "1",
        "coarse_enough": "333"
    },
    "solver": {
        "type": "lgmres",
        "tol": "9.9999999999999995e-07",
        "maxiter": "1000"
    }
}
[WARNING] AMGCL: Unknown parameter nullspace
AMGCL Linear Solver: Iterations: 52
Error: 4.17743e-07
~~~

with the warning suggesting that it is most probably NOT using the nullspace.

If i run in OpenMP mode, i get 101 iterations WITH the nullspace (provde_coordinates=true), and 47 without

~~~
{
    "precond": {
        "class": "amg",
        "relax": {
            "type": "ilu0"
        },
        "coarsening": {
            "type": "smoothed_aggregation",
            "aggr": {
                "eps_strong": "0",
                "block_size": "3"
            },
            "nullspace": {
                "cols": "6",
                "rows": "15120",
                "B": "0x296bb10"
            }
        },
        "npre": "1",
        "npost": "1",
        "coarse_enough": "333"
    },
    "solver": {
        "type": "lgmres",
        "tol": "9.9999999999999995e-07",
        "maxiter": "1000"
    }
}
AMGCL Memory Occupation : 93.88 M
AMGCL Linear Solver: Iterations: 101
Error: 3.18788e-07
~~~

this suggests there is still something wrong with the nullspace prescription

**Comment by ddemidov** [comment]
1. nullspace is not yet supported in amgcl mpi, only in serial version.
2. can you send me the matrix and the coordinate file?

**Comment by ddemidov** [comment]
@RiccardoRossi, here is what I get with your example:

**Not using rigid body modes, use blocked aggregation with block_size=3:**

```
./solver -A problems/coordinates/A.mm -f problems/coordinates/b.mm.rhs \
  -p precond.coarsening.aggr.eps_strong=0 precond.coarsening.aggr.block_size=3
Solver
======
Type:             BiCGStab
Unknowns:         15120
Memory footprint: 826.88 K

Preconditioner
==============
Number of levels:    2
Operator complexity: 1.13
Grid complexity:     1.06
Memory footprint:    16.27 M

level     unknowns       nonzeros      memory
---------------------------------------------
    0        15120         605808     15.11 M (88.63%)
    1          846          77688      1.16 M (11.37%)

Iterations: 70
Error:      1.40754e-09

[Profile:       1.032 s] (100.00%)
[  reading:     0.560 s] ( 54.22%)
[  setup:       0.038 s] (  3.71%)
[  solve:       0.434 s] ( 42.02%)
```

**Using rigid body modes:**

```
./solver -A problems/coordinates/A.mm -f problems/coordinates/b.mm.rhs \
  -p precond.coarsening.aggr.eps_strong=0 -C problems/coordinates/C.mm
Solver
======
Type:             BiCGStab
Unknowns:         15120
Memory footprint: 826.88 K

Preconditioner
==============
Number of levels:    2
Operator complexity: 1.51
Grid complexity:     1.11
Memory footprint:    24.90 M

level     unknowns       nonzeros      memory
---------------------------------------------
    0        15120         605808     20.30 M (66.09%)
    1         1692         310896      4.61 M (33.91%)

Iterations: 15
Error:      8.89349e-09

[Profile:       0.846 s] (100.00%)
[  reading:     0.558 s] ( 65.89%)
[  setup:       0.139 s] ( 16.46%)
[  solve:       0.149 s] ( 17.59%)
```

I think the problem in Kratos is that `block_size` is set even when rigid body modes are provided.
Can you please check if 20043b9228 fixes the issue?

**Comment by RiccardoRossi** [comment]
Hi @ddemidov unfortunately your fix doesn't appear to work...

~~~json
AMGCL Linear Solver: mndof: 3
{
    "precond": {
        "class": "amg",
        "relax": {
            "type": "ilu0"
        },
        "coarsening": {
            "type": "smoothed_aggregation",
            "aggr": {
                "eps_strong": "0"
            },
            "nullspace": {
                "cols": "6",
                "rows": "15120",
                "B": "0x22dd1f0"
            }
        },
        "npre": "1",
        "npost": "1",
        "coarse_enough": "333"
    },
    "solver": {
        "type": "lgmres",
        "tol": "9.9999999999999995e-07",
        "maxiter": "1000"
    }
}
AMGCL Memory Occupation : 93.48 M
AMGCL Linear Solver: Iterations: 101
Error: 3.18788e-07

~~~

**Comment by RiccardoRossi** [comment]
so with the latest commit this works. nevertheless i do not think this is a good fix.

i think that instead we should alsways use the ScalarSolve (even if internally, the block_size to be used in aggregation is >1) whenever the variable "mUseBlockMatricesIfPossible" is set to false.

the problem is that i don't know how to pass that variable to the amgcl_solver_impl

@ddemidov what do you think?

**Comment by ddemidov** [comment]
> i think that instead we should alsways use the ScalarSolve (even if internally, the block_size to be used in aggregation is >1) whenever the variable "mUseBlockMatricesIfPossible" is set to false.

@RiccardoRossi, I think 2e710ad should do this. Note that I rebased on latest master, so if you want to test the changes in the `amgcl-fix-rbm` branch, you should
```
git fetch
git reset --hard origin/amgcl-fix-rbm
```

**Comment by RiccardoRossi** [comment]
@ddemidov i will accept the PR as it is,
nevertheless the problem i foresee is the following (already present, so this doesn't make the situation worst),

imagine  that the dofs are
u1x u1y u1z u2x u2y

where u1x means displacement of node 1 component x

this may happen in kratos if one uses the so called EliminationBuilderAndSolver

the kratos would attempt to detect automatically the block size, however as here we have 3dofs for node1 and 2dofs for node2 we would conclude that there is not a common block size used through the dofset, and conclude mBlockSize=1.

ideally however we would provide 3 rigid body moves in this context (although the one for node2 would be truncated since the dof u2z is not there.)

note sure what would happen in that case...

**Review by RiccardoRossi — APPROVED** [review]
thx a lot for these corrections!!

**Comment by ddemidov** [comment]
@RiccardoRossi,

> nevertheless the problem i foresee is the following...

This would not work before the PR either. I think to make it work we could provide a way to set the near null-space vectors directly (as opposed to providing the coordinates), but we can always implement this later.

**Comment by philbucher** [comment]
Hi @ddemidov 
I am getting some compiler warnings (I think since this merge):
~~~
In file included from /home/philippb/software/Kratos_prod/external_libraries/amgcl/relaxation/runtime.hpp:50:0,
                 from /home/philippb/software/Kratos_prod/kratos/linear_solvers/amgcl_solver_impl.cpp:11:
/home/philippb/software/Kratos_prod/external_libraries/amgcl/relaxation/chebyshev.hpp: In member function ‘void amgcl::relaxation::chebyshev<Backend>::solve(const Matrix&, const VectorB&, VectorX&) const [with Matrix = amgcl::backend::crs<double, long int, long int>; VectorB = amgcl::backend::numa_vector<double>; VectorX = amgcl::backend::numa_vector<double>; Backend = amgcl::backend::builtin<double>]’:
/home/philippb/software/Kratos_prod/external_libraries/amgcl/relaxation/chebyshev.hpp:181:25: warning: ‘alpha’ may be used uninitialized in this function [-Wmaybe-uninitialized]
             scalar_type alpha, beta;
                         ^
/home/philippb/software/Kratos_prod/external_libraries/amgcl/relaxation/chebyshev.hpp: In member function ‘void amgcl::relaxation::chebyshev<Backend>::solve(const Matrix&, const VectorB&, VectorX&) const [with Matrix = amgcl::backend::crs<amgcl::static_matrix<double, 2, 2>, long int, long int>; VectorB = amgcl::backend::numa_vector<amgcl::static_matrix<double, 2, 1> >; VectorX = amgcl::backend::numa_vector<amgcl::static_matrix<double, 2, 1> >; Backend = amgcl::backend::builtin<amgcl::static_matrix<double, 2, 2> >]’:
/home/philippb/software/Kratos_prod/external_libraries/amgcl/relaxation/chebyshev.hpp:181:25: warning: ‘alpha’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/philippb/software/Kratos_prod/external_libraries/amgcl/relaxation/chebyshev.hpp: In member function ‘void amgcl::relaxation::chebyshev<Backend>::solve(const Matrix&, const VectorB&, VectorX&) const [with Matrix = amgcl::backend::crs<amgcl::static_matrix<double, 3, 3>, long int, long int>; VectorB = amgcl::backend::numa_vector<amgcl::static_matrix<double, 3, 1> >; VectorX = amgcl::backend::numa_vector<amgcl::static_matrix<double, 3, 1> >; Backend = amgcl::backend::builtin<amgcl::static_matrix<double, 3, 3> >]’:
/home/philippb/software/Kratos_prod/external_libraries/amgcl/relaxation/chebyshev.hpp:181:25: warning: ‘alpha’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/philippb/software/Kratos_prod/external_libraries/amgcl/relaxation/chebyshev.hpp: In member function ‘void amgcl::relaxation::chebyshev<Backend>::solve(const Matrix&, const VectorB&, VectorX&) const [with Matrix = amgcl::backend::crs<amgcl::static_matrix<double, 4, 4>, long int, long int>; VectorB = amgcl::backend::numa_vector<amgcl::static_matrix<double, 4, 1> >; VectorX = amgcl::backend::numa_vector<amgcl::static_matrix<double, 4, 1> >; Backend = amgcl::backend::builtin<amgcl::static_matrix<double, 4, 4> >]’:
/home/philippb/software/Kratos_prod/external_libraries/amgcl/relaxation/chebyshev.hpp:181:25: warning: ‘alpha’ may be used uninitialized in this function [-Wmaybe-uninitialized]
~~~

can you please have a look when you have time?
I am using GCC 5.4

**Comment by ddemidov** [comment]
This is a false warning, but #6303 should fix it.

**Comment by ddiezrod** [comment]
Hi @ddemidov and @RiccardoRossi,  One of our Filling QA tests in @KratosMultiphysics/altair side is failing after merging this PR. The difference is very small in the residual but it builds up until it gets noticeable. 

I cannot tell which result is better (before or after this PR) so I can just modify the QA results so it runs ok , but I just wanted to make sure if you expected this small change in results or it might be a silent bug.

**Comment by ddemidov** [comment]
Is it possible to get the system matrix and the RHS for the test (one of the time steps) so I could reproduce it?

**Comment by ddiezrod** [comment]
[amgcl_matrices.zip](https://github.com/KratosMultiphysics/Kratos/files/4171176/amgcl_matrices.zip)
@ddemidov There you go! Thanks, I will wait for your answer.

**Comment by ddemidov** [comment]
Does the archive also contain amgcl configuration used in the test?

**Comment by ddiezrod** [comment]
Nope, sorry:
"linear_solver_settings": {
                "preconditioner_type"            : "amg",
                "solver_type"                    : "amgcl",
                "smoother_type"                  :"ilu0",
                "krylov_type"                    : "lgmres",
                "coarsening_type"                : "aggregation",
                "max_iteration"                  : 400,
                "provide_coordinates"            : false,
                "gmres_krylov_space_dimension"   : 50,
                "verbosity"                      : 0,
                "tolerance"                      : 1e-5,
                "scaling"                        : true,
                "block_size"                     : 1,
                "use_block_matrices_if_possible" : false,
                "coarse_enough"                  : 5000,
                "max_levels"                     : -1,
                "pre_sweeps"                     : 1,
                "post_sweeps"                    : 1,
                "use_gpgpu"                      : false
            }

Is this enough?

**Comment by RiccardoRossi** [comment]
dani, best of all use verbosity=4 

that will output the internal settings of amgcl, which Denis can use automatically

**Comment by ddemidov** [comment]
I am getting this with these settings:
```
./solver -A kratos-6077/A_0_1.mm -f kratos-6077/b_0_1.mm.rhs solver.type=lgmres solver.maxiter=400 solver.M=50 solver.tol=1e-5 precond.relax.type=ilu0 precond.coarsening.type=aggregation
System matrix
=============
Memory footprint: 2.11 M

Solver
======
Type:             LGMRES(50,3)
Unknowns:         2976
Memory footprint: 1.36 M

Preconditioner
==============
Number of levels:    1
Operator complexity: 1.00
Grid complexity:     1.00
Memory footprint:    12.00 M

level     unknowns       nonzeros      memory
---------------------------------------------
    0         2976         136864     12.00 M (100.00%)

Iterations: 1
Error:      3.04387e-14

[Profile:       0.319 s] (100.00%)
[ self:         0.010 s] (  3.00%)
[  reading:     0.092 s] ( 28.86%)
[  setup:       0.214 s] ( 67.15%)
[  solve:       0.003 s] (  0.99%)
```
does it seem correct? The system is so small that a direct solver is used as a preconditioner

**Comment by ddemidov** [comment]
So, `git bisect` tells me that the only commit (starting with 100 commits
behind the current amgcl master) affecting the reported residual is
ddemidov/amgcl@37dcf25a46a7eab2c3bbd06359df13163c6d5b41 (makes sense to me).

Before that, the result was:

```
Solver
======
Type:             LGMRES(50,3)
Unknowns:         2976
Memory footprint: 1.36 M

Preconditioner
==============
Number of levels:    1
Operator complexity: 1.00
Grid complexity:     1.00
Memory footprint:    14.11 M

level     unknowns       nonzeros      memory
---------------------------------------------
    0         2976         136864     14.11 M (100.00%)

Iterations: 1
Error:      1.72256e-14

[Profile:       0.222 s] (100.00%)
[ self:         0.000 s] (  0.11%)
[  reading:     0.088 s] ( 39.78%)
[  setup:       0.130 s] ( 58.86%)
[  solve:       0.003 s] (  1.25%)
```

After the commit, the residual is stable and is equal to `3.04387e-14` as reported in the comment above. Are you sure you really are using these amgcl options (specifically, is `coarse_enough` really is equal to 5000)?

**Comment by ddiezrod** [comment]
This si strange but it seems we are getting different results. For me, in previous verson (exactly pointing to commit 55728c50 (19th December)

```
AMGCL Linear Solver: mndof: 4
{
    "precond": {
        "class": "amg",
        "relax": {
            "type": "ilu0"
        },
        "coarsening": {
            "type": "aggregation",
            "aggr": {
                "eps_strong": "0",
                "block_size": "4"
            }
        },
        "npre": "1",
        "npost": "1",
        "coarse_enough": "1250"
    },
    "solver": {
        "type": "lgmres",
        "tol": "1.0000000000000001e-05",
        "maxiter": "400"
    }
}
AMGCL Memory Occupation : 14.90 M
AMGCL Linear Solver: Iterations: 1
Error: 1.83682e-14

Convergence Criteria:  VELOC.: ratio = 1; exp.ratio = 0.03 abs = 0.00375475 exp.abs = 1e-06
Convergence Criteria:  PRESS.: ratio = 1; exp.ratio = 0.03 abs = 0.557044 exp.abs = 1e-06
```

and the new one, pointing to 3ce3c203 (6th February)
```

AMGCL Linear Solver: mndof: 4
{
    "precond": {
        "class": "amg",
        "relax": {
            "type": "ilu0"
        },
        "coarsening": {
            "type": "aggregation",
            "aggr": {
                "eps_strong": "0",
                "block_size": "4"
            }
        },
        "npre": "1",
        "npost": "1",
        "coarse_enough": "1250"
    },
    "solver": {
        "type": "lgmres",
        "tol": "1.0000000000000001e-05",
        "maxiter": "400"
    }
}
AMGCL Memory Occupation : 3.80 M
AMGCL Linear Solver: Iterations: 24
Error: 5.74448e-06

Convergence Criteria:  VELOC.: ratio = 1; exp.ratio = 0.03 abs = 0.00375478 exp.abs = 1e-06
Convergence Criteria:  PRESS.: ratio = 1; exp.ratio = 0.03 abs = 0.557047 exp.abs = 1e-06
```

It surprises me the difference in number of iterations and the memory allocation.

**Comment by ddemidov** [comment]
The amgcl parameters here are a bit different (block_size and coarse_enough). I'll test with these.

**Comment by ddiezrod** [comment]
@ddemidov Thats true, I called the python linear solver factory with the first parameters, but somehow they are not the final ones that are being used. I will check this also.

**Comment by ddemidov** [comment]
Looks like in your first case you get a single-level hierarchy (solved with a direct solver in single iteration), while in the second case you get at least two levels, hence the iterations up to the required tolerance (1e-5).

**Comment by ddemidov** [comment]
Could this change be the reason?

https://github.com/KratosMultiphysics/Kratos/pull/6077#issuecomment-575189991, https://github.com/KratosMultiphysics/Kratos/commit/2e710add7f5a5578eaf58e16ab51f6615e05f961

**Comment by ddemidov** [comment]
@ddiezrod, Can you set `use_block_matrices_if_possible=true` to test this?

**Comment by ddiezrod** [comment]
@ddemidov That was it! Setting it to true, old behaviour is recovered and QA passes fine. Do you think I should leave it as true?

**Comment by ddemidov** [comment]
If you need the old behavior, then yes. But I think both options are correct from amgcl point of view.
@RiccardoRossi , what do you think?

EDIT: just to note: the 'old' behavior only worked in the way it worked because the matrix was small enough for amgcl to consider it 'coarse enough' to be solved directly. If you increase the matrix size beyond 5000 dofs, then the test will fail again. So may be the criteria is too tight for the tolerance you ask from amgcl.

**Comment by ddiezrod** [comment]
I understand, my concern was that there was some kind of silent bug behind. I can update the QA resutls or modify the parameter. I do not even know what the "block_size" is so whatever you guys think its the best.

**Comment by RiccardoRossi** [comment]
The point is that using "use_block_matrices_if_possible=true" should be faster, but at the price of making a copy of the matrix (so more memory).

leaving it to false allows more options and should use a tad less memory, however it should be slightly slower.

before this PR in some corner case the option was silently ignored, which explains the change in behaviour. That is, the one we have now is the "correct one" however it does change the previous behaviour (sorry, i didn't realize it would affect)

**Comment by ddiezrod** [comment]
Ok @RiccardoRossi ! I will update the test results then, Thanks!

**Pull request merged**
